### ISSUE TITLE
Switching to text driven programming

### ISSUE THREAD

**Issue by neslihanturan** [commons-app/apps-android-commons#1883]
**Summary:** 

Hi team, since our time spent for bug fixes seems a little high, maybe switching to text driven programming can be a solution. Since on our WMCZ grant session, we will have several quality improvement, then our code base may be ready for text driven programming. What do you think?

**Comment by misaochan** [comment]
I am assuming you meant "test-driven programming", so I took the liberty of editing your post. ;)

I agree, I think this would be well worth the investment. Would love comments from experienced devs especially - @nicolas-raoul @whym @dbrant @psh ?

**Comment by nicolas-raoul** [comment]
I don't think test-driven programming is a great fit for most parts of our project because:
- Most bugs are in parts that unit tests can not reach (or at least can not reach without a really big effort), such as the GUI or 2FA.
- In contrast with most software (most software is developed for companies, for instance a company's billing system or a factory's machine software), we have the chance to have a community of testers. Test-driven is a great way to reduce the number of bugs in the output of your IDE, but it comes at a great cost. Companies are more willing to afford it, because faulty software could result in destroying a factory machine, and they have no community to get feedback from.

That being said, there are parts of the project that could certainly benefit from test-driven programming, so I suggest starting with those. For instance the non-GUI part of EXIF parsing and EXIF modification, this could be a good start.

**Comment by neslihanturan** [comment]
If we can well define where we want to have it, we can include this as a rule to contributors documents. So that new comers also can now, they need to write tests if possible. 

And since we can't write tests for GUI, maybe we can make sure we separated GUI layer as much as we can.

**Comment by whym** [comment]
In general, I think it's a good exercise to ask yourself 'can I write tests?' when creating and modifying a feature, and perhaps more importantly, when fixing a bug so that the same bug will not happen again easily.

We have utility classes and methods (e.g. app/src/main/java/fr/free/nrw/commons/Utils.java) for which we can naturally write tests. Perhaps we can start from those.

GUI features are harder to test, but not impossible in my opinion. I believe something like app/src/androidTest/java/fr/free/nrw/commons/SettingsActivityTest.java is a good way to catch bugs that might sneak in unintentionally. High coverage would probably be unrealistic, but a few GUI tests will be great to have to prevent breaking things.

**Comment by ilgazer** [comment]
In my opinion, having more tests could be beneficial. On the other hand, making tests a requirement could dissuade less experienced developers(that's me) from developing new features. For example, even though I built upon psh's MVP architecture for #1796, I would have shied away from working on that if I had to write tests for my code, something I have no experience or knowledge about. The same holds for my EXIF modification code. In conclusion, I think that while encouraging tests is a good idea, mandatory tests would do more harm than good.

**Comment by ashishkumar468** [comment]
Although TDD enables refactoring in an anxiety-free, predictable way, and of-course enforces standards, reaching slowly towards bug free and more stabler product, not all of the developers(actually most of them) follow it and actually know it. Even though it has a lot of advantages, we all know that it would require extra efforts to those who are not acquainted with it. Having it would be great, but might discourage early stage contributors. And if we talk about improvements in code standards improvement, of-course TDD would have been the best solution, but for the time period, until the app reaches stable in terms of development and adding new features (which i am not sure it is now), where we have enough of resources and time and not loaded with new features, IMO, and I may be wrong, we can go ahead with the current policy of not mandating it (but of-course encouraging it to those who can and are willing to go for it). Also code standards can and is being currently enforced by the code reviewers (thanks to the concept of PR's). But not accepting PR's because of test cases, in my opinion, would not be the best thing to do as of now. Although I would recommend to introduce a code architecture pattern, say MVP or MVVM (Please correct me if we already follow any of these). That will automatically raise code standards and improve code quality.

**Comment by ilgazer** [comment]
@misaochan Your compromise of only requiring grantees to write unit tests is a really good idea. If we will apply as a GSoC project, maybe we could have the same requirement for them.

**Comment by misaochan** [comment]
@ilgazer Good idea! 👍

**Comment by ilgazer** [comment]
Also, I think linking to some good resources on where to learn testing would help a lot. I distinctly remember being overwhelmed by the abundance of stuff to learn and giving up when I tried to start writing tests for #1712. I still have no clue about how to write tests.

**Comment by NotSoImportant** [comment]
Is this a stale issue?

**Comment by neslihanturan** [comment]
I don't think this is gonna happen soon. If you have suggestions/things to say about this issue feel free to share.

**Comment by NotSoImportant** [comment]
I was just wondering as something was put to done if now you have applied it or what are the next steps.

**Comment by misaochan** [comment]
I would consider this a semi-success. All PRs for new features/enhancements now require unit tests, as mentioned in our contribution guidelines. However, I don't think most contributors are applying true TDD principles (which is different from just writing tests for their code).
